File: src/lib/engine/external-context.ts
// External market context — pure types + normalizers shared by the worker
// ingester, the server-side assembler, the API payload, and the client hook.
// Framework-free and network-free so every transform is fixture-testable.
//
// This is the news/context plane, NOT the signal engine: nothing here feeds
// `EvaluateInput` or any deterministic verdict path (ENGINE_VERSION untouched).
// The assembled context is secondary evidence for the AI analyst prompt only.

/** Normalized CoinGecko `/global` poll — one persisted breadth snapshot. */
export interface MarketContextSnapshotInput {
  totalMcapUsd: number;
  btcDominance: number;
  ethDominance: number | null;
  mcapChange24hPct: number | null;
  source: string;
}

/**
 * Parse a CoinGecko `/global` response body into a snapshot input.
 * Returns null when the payload doesn't carry the fields we need — a schema
 * drift upstream must surface as an ingest error, never as a NaN row.
 */
export function normalizeCoinGeckoGlobal(payload: unknown): MarketContextSnapshotInput | null {
  const data = (payload as { data?: Record<string, unknown> } | null)?.data;
  if (!data || typeof data !== "object") return null;
  const mcap = (data.total_market_cap as Record<string, unknown> | undefined)?.usd;
  const pct = data.market_cap_percentage as Record<string, unknown> | undefined;
  const btc = pct?.btc;
  if (typeof mcap !== "number" || !Number.isFinite(mcap) || mcap <= 0) return null;
  if (typeof btc !== "number" || !Number.isFinite(btc) || btc <= 0) return null;
  const eth = pct?.eth;
  const change = data.market_cap_change_percentage_24h_usd;
  return {
    totalMcapUsd: mcap,
    btcDominance: btc,
    ethDominance: typeof eth === "number" && Number.isFinite(eth) ? eth : null,
    mcapChange24hPct: typeof change === "number" && Number.isFinite(change) ? change : null,
    source: "coingecko",
  };
}

// ── Assembled context (the /api/external-context payload) ───────────────────

/** Provenance + freshness stamp carried by every populated section. */
export interface ContextProvenance {
  source: string;
  /** When the underlying data was fetched/computed (ISO). */
  asOf: string;
  /** Older than the section's freshness threshold — usable, but must be labeled. */
  stale: boolean;
}

/**
 * Why a section is absent or suspect. `missing` is an expected gap (e.g. an
 * out-of-universe symbol has no relative read); `unconfigured` means no API
 * key; `error` means a configured pipeline failed and should alarm.
 */
export interface SectionStatus {
  section: "breadth" | "relative" | "recentCatalysts" | "recentHighImpact" | "upcoming" | "social";
  status: "ok" | "stale" | "missing" | "unconfigured" | "error";
  reason?: string;
  asOf?: string;
}

/** Market-wide backdrop: is the tape risk-on, and is money in or out of BTC? */
export interface BreadthContext {
  /** e.g. "Risk On" — the snapshot's BTC-daily regime read; null when only the provider row is available. */
  btcRegime: string | null;
  btcChange24hPct: number | null;
  btcDominancePct: number | null;
  btcDominanceDelta24hPp: number | null; // percentage-point change vs ~24h ago
  totalMcapUsd: number | null;
  mcapChange24hPct: number | null;
  fearGreed: number | null;
  fearGreedLabel: string | null;
  provenance: ContextProvenance;
}

/** Viewed asset vs BTC: is this move the asset's own, or just the market's? */
export interface RelativeContext {
  rsBtc24hPp: number;
  rsBtc7dPp: number;
  corrBtc7d: number | null;
  /** Asset/BTC price ratio from {ASSET}USDT / BTCUSDT closes. */
  ratio: number | null;
  ratioChange7dPct: number | null;
  provenance: ContextProvenance;
}

/** One recent confirmed event (token_event row projected for the context payload). */
export interface ContextEventItem {
  kind: string;
  severity: string;
  title: string;
  source: string;
  url: string | null;
  publishedAt: string;
}

/** One upcoming calendar catalyst (catalyst_event row projection). */
export interface UpcomingCatalystItem {
  symbol: string;
  kind: string;
  title: string;
  occursAt: string;
  source: string;
  url: string | null;
  /** null = size unknown (free-tier norm) — a scheduling fact, not a supply-pressure signal. */
  percentOfSupply: number | null;
  votes: number | null;
  confidencePct: number | null;
}

export interface ExternalContext {
  symbol: string;
  assembledAt: string;
  breadth: BreadthContext | null;
  relative: RelativeContext | null;
  /** ≤48h back — plausible contributors to the current move. */
  recentCatalysts: ContextEventItem[] | null;
  /** ≤7d back, severity-gated — standing high-impact context. */
  recentHighImpact: ContextEventItem[] | null;
  /** ≤7d ahead for this symbol. */
  upcoming: UpcomingCatalystItem[] | null;
  /** ≤7d ahead, market-wide (BTC/ETH/MARKET). */
  marketEvents: UpcomingCatalystItem[] | null;
  /** Reserved slot — LunarCrush integration deferred until limits verified. */
  social: null;
  /** Always present: per-section status, so degradation is observable, not silent. */
  degradation: SectionStatus[];
}

// ── Freshness thresholds (shared by service + health view) ──────────────────

export const BREADTH_STALE_MS = 60 * 60_000; // global snapshot older than 1h
export const CALENDAR_STALE_MS = 12 * 60 * 60_000; // calendar last-ok older than 12h
export const RSS_STALE_MS = 60 * 60_000;

export function fearGreedLabel(value: number): string {
  if (value >= 75) return "Extreme Greed";
  if (value >= 55) return "Greed";
  if (value > 45) return "Neutral";
  if (value > 25) return "Fear";
  return "Extreme Fear";
}

File: engine/smc/external_context.py
"""External market context — pure normalizers for the worker's breadth
ingestion (port of external-context.ts's ingester plane).

This is the news/context plane, NOT the signal engine: nothing here feeds
``EvaluateInput`` or any deterministic verdict path (ENGINE_VERSION
untouched). The assembled-context payload (/api/external-context) stays a
legacy-web read model over the same Postgres rows.
"""

from __future__ import annotations

import math
from dataclasses import dataclass


@dataclass(slots=True)
class MarketContextSnapshotInput:
    """Normalized global-breadth poll (CoinGecko or CoinMarketCap) — one
    persisted snapshot."""

    total_mcap_usd: float
    btc_dominance: float
    eth_dominance: float | None
    mcap_change_24h_pct: float | None
    source: str


def _finite_number(value: object) -> float | None:
    if isinstance(value, bool):
        return None
    if isinstance(value, (int, float)) and math.isfinite(value):
        return float(value)
    return None


def normalize_coingecko_global(payload: object) -> MarketContextSnapshotInput | None:
    """Parse a CoinGecko ``/global`` response body into a snapshot input.
    Returns None when the payload doesn't carry the fields we need — a schema
    drift upstream must surface as an ingest error, never as a NaN row."""
    data = payload.get("data") if isinstance(payload, dict) else None
    if not isinstance(data, dict):
        return None
    total_mcap = data.get("total_market_cap")
    mcap = _finite_number(total_mcap.get("usd")) if isinstance(total_mcap, dict) else None
    pct = data.get("market_cap_percentage")
    btc = _finite_number(pct.get("btc")) if isinstance(pct, dict) else None
    if mcap is None or mcap <= 0 or btc is None or btc <= 0:
        return None
    eth = _finite_number(pct.get("eth")) if isinstance(pct, dict) else None
    change = _finite_number(data.get("market_cap_change_percentage_24h_usd"))
    return MarketContextSnapshotInput(
        total_mcap_usd=mcap,
        btc_dominance=btc,
        eth_dominance=eth,
        mcap_change_24h_pct=change,
        source="coingecko",
    )


def normalize_coinmarketcap_global(payload: object) -> MarketContextSnapshotInput | None:
    """Parse a CoinMarketCap ``/v1/global-metrics/quotes/latest`` response
    into the same snapshot shape. Available on the free Basic plan (1
    credit/call), so a working CMC key fully substitutes for CoinGecko as the
    breadth source."""
    data = payload.get("data") if isinstance(payload, dict) else None
    if not isinstance(data, dict):
        return None
    quote_wrap = data.get("quote")
    quote = quote_wrap.get("USD") if isinstance(quote_wrap, dict) else None
    mcap = _finite_number(quote.get("total_market_cap")) if isinstance(quote, dict) else None
    btc = _finite_number(data.get("btc_dominance"))
    if mcap is None or mcap <= 0 or btc is None or btc <= 0:
        return None
    eth = _finite_number(data.get("eth_dominance"))
    change = (
        _finite_number(quote.get("total_market_cap_yesterday_percentage_change"))
        if isinstance(quote, dict)
        else None
    )
    return MarketContextSnapshotInput(
        total_mcap_usd=mcap,
        btc_dominance=btc,
        eth_dominance=eth,
        mcap_change_24h_pct=change,
        source="coinmarketcap",
    )


def fear_greed_label(value: float) -> str:
    """The alternative.me index bands."""
    if value >= 75:
        return "Extreme Greed"
    if value >= 55:
        return "Greed"
    if value > 45:
        return "Neutral"
    if value > 25:
        return "Fear"
    return "Extreme Fear"
